core/node/node_sync/src/batch_transaction_updater/mod.rs

```diff
@@ -0,0 +1,314 @@
+
+        Ok(SLBlockNumbers {
+            finalized,
+            fast_finality,
+        })
+    }
+
+    async fn apply_status_update(
+        &mut self,
+        db_eth_history_id: u32,
+        receipt: TransactionReceipt,
+        sl_block_numbers: &SLBlockNumbers,
+    ) -> anyhow::Result<()> {
+        let updated_status =
+            sl_block_numbers.get_finality_status_for_block(receipt.block_number.unwrap());
+
+        let mut connection = self
+            .pool
+            .connection_tagged("batch_transaction_updater")
+            .await?;
+
+        let db_eth_history_tx = connection
+            .eth_sender_dal()
+            .get_eth_tx_history_by_id(db_eth_history_id)
+            .await?;
+
+        let batch_number = connection
+            .blocks_dal()
+            .get_l1_batch_number_by_eth_tx_id(db_eth_history_tx.eth_tx_id)
+            .await?
+            .expect("eth_tx_history row must match a l1 batch");
+
+        // validate the transaction against db
+        let () = match db_eth_history_tx.tx_type {
+            AggregatedActionType::Commit => {
+                self.l1_transaction_verifier
+                    .validate_commit_tx(receipt, batch_number)
+                    .await?
+            }
+            AggregatedActionType::PublishProofOnchain => {
+                self.l1_transaction_verifier
+                    .validate_prove_tx(receipt, batch_number)
+                    .await?
+            }
+            AggregatedActionType::Execute => {
+                self.l1_transaction_verifier
+                    .validate_execute_tx(receipt, batch_number)
+                    .await?
+            }
+        };
+
+        connection
+            .eth_sender_dal()
+            .mark_received_eth_tx_as_verified(db_eth_history_id, updated_status)
+            .await?;
+        Ok(())
+    }
+
+    async fn update_statuses(&mut self, sl_block_numbers: SLBlockNumbers) -> anyhow::Result<i32> {
+        let mut updated_count = 0;
+        for transaction_type in TRANSACTION_TYPES {
+            for finality_status in UPDATABLE_FINALITY_STATUS {
+                let to_process_entry =
+                    &mut self.next_to_process[finality_status as usize][transaction_type as usize];
+
+                // if not cached, try to load from DB
+                if to_process_entry.is_none() {
+                    let next_to_process = self
+                        .pool
+                        .connection_tagged("batch_transaction_updater")
+                        .await?
+                        .eth_sender_dal()
+                        .get_oldest_tx_by_status_and_type(finality_status, transaction_type)
+                        .await?;
+                    *to_process_entry = next_to_process.map(|tx| NextToProcess {
+                        tx_hash: tx.tx_hash,
+                        db_eth_history_id: tx.id,
+                        seen_at_block: None,
+                    });
+                }
```

Comment:
Why not simply, 
```
            SELECT
                eth_txs.*
            FROM
                eth_txs
            JOIN eth_txs_history ON eth_txs.confirmed_eth_tx_history_id = eth_txs_history.id
            WHERE
                AND eth_txs_history.finality_status != 'finalized'
            ORDER BY
                eth_txs.id

```

Without statuses and so on

Reply:
BatchStatusUpdater inserts transactions of the same type in sequence. I think it cannot do much better. We want to get from DB the transaction(s) that are closest to be finalized/fast-finalized. I have implemented it like this to get the 6 candidates for updates. If I would go with your suggestion, I think I would need to check all pending transactions to find the closest one for status change.